### ISSUE TITLE
Crash: when accessing 2nd mocked class object (see example) 

### ISSUE THREAD

**Issue by mohand1993** [moq/moq4#449]
The following code block cause the crash below.
```

var obj = new Mock<Object>().Object;
var stream = new Mock<Stream>().Object;
```

> * Assertion at dynamic-image.c:209, condition `prev == MONO_HANDLE_RAW (obj)' not met
> 
> Stacktrace:
> 
>   at <unknown> <0xffffffff>
>   at (wrapper managed-to-native) System.Reflection.Emit.ModuleBuilder.getToken (System.Reflection.Emit.ModuleBuilder,object,bool) <0x00015>
>   at System.Reflection.Emit.ModuleBuilder.GetToken (System.Reflection.MemberInfo,bool) [0x00061] in /Users/builder/data/lanes/4992/mono-mac-sdk/external/bockbuild/builds/mono-x64/mcs/class/corlib/System.Reflection.Emit/ModuleBuilder.cs:793
>   at System.Reflection.Emit.ModuleBuilderTokenGenerator.GetToken (System.Reflection.MemberInfo,bool) [0x00000] in /Users/builder/data/lanes/4992/mono-mac-sdk/external/bockbuild/builds/mono-x64/mcs/class/corlib/System.Reflection.Emit/ModuleBuilder.cs:1207
>   at System.Reflection.Emit.ILGenerator.Emit (System.Reflection.Emit.OpCode,System.Reflection.MethodInfo) [0x0004e] in /Users/builder/data/lanes/4992/mono-mac-sdk/external/bockbuild/builds/mono-x64/mcs/class/corlib/System.Reflection.Emit/ILGenerator.cs:734
>   at Castle.DynamicProxy.Generators.Emitters.SimpleAST.MethodTokenExpression.Emit (Castle.DynamicProxy.Generators.Emitters.IMemberEmitter,System.Reflection.Emit.ILGenerator) [0x0000c] in <1da8a528a77240358e1afd90a17b1a27>:0
>   at Castle.DynamicProxy.Generators.Emitters.SimpleAST.AssignStatement.Emit (Castle.DynamicProxy.Generators.Emitters.IMemberEmitter,System.Reflection.Emit.ILGenerator) [0x00019] in <1da8a528a77240358e1afd90a17b1a27>:0
>   at Castle.DynamicProxy.Generators.Emitters.CodeBuilders.AbstractCodeBuilder.Generate (Castle.DynamicProxy.Generators.Emitters.IMemberEmitter,System.Reflection.Emit.ILGenerator) [0x00058] in <1da8a528a77240358e1afd90a17b1a27>:0
>   at Castle.DynamicProxy.Generators.Emitters.ConstructorEmitter.Generate () [0x0001e] in <1da8a528a77240358e1afd90a17b1a27>:0
>   at Castle.DynamicProxy.Generators.Emitters.AbstractTypeEmitter.EnsureBuildersAreInAValidState () [0x000c6] in <1da8a528a77240358e1afd90a17b1a27>:0
>   at Castle.DynamicProxy.Generators.Emitters.AbstractTypeEmitter.BuildType () [0x00001] in <1da8a528a77240358e1afd90a17b1a27>:0
>   at Castle.DynamicProxy.Generators.ClassProxyGenerator.GenerateType (string,System.Type[],Castle.DynamicProxy.Generators.INamingScope) [0x0015d] in <1da8a528a77240358e1afd90a17b1a27>:0
>   at Castle.DynamicProxy.Generators.ClassProxyGenerator/<>c__DisplayClass1.<GenerateCode>b__0 (string,Castle.DynamicProxy.Generators.INamingScope) [0x0000e] in <1da8a528a77240358e1afd90a17b1a27>:0
>   at Castle.DynamicProxy.Generators.BaseProxyGenerator.ObtainProxyType (Castle.DynamicProxy.Generators.CacheKey,System.Func`3<string, Castle.DynamicProxy.Generators.INamingScope, System.Type>) [0x00141] in <1da8a528a77240358e1afd90a17b1a27>:0
>   at Castle.DynamicProxy.Generators.ClassProxyGenerator.GenerateCode (System.Type[],Castle.DynamicProxy.ProxyGenerationOptions) [0x00064] in <1da8a528a77240358e1afd90a17b1a27>:0
>   at Castle.DynamicProxy.DefaultProxyBuilder.CreateClassProxyType (System.Type,System.Type[],Castle.DynamicProxy.ProxyGenerationOptions) [0x0002c] in <1da8a528a77240358e1afd90a17b1a27>:0
>   at Castle.DynamicProxy.ProxyGenerator.CreateClassProxyType (System.Type,System.Type[],Castle.DynamicProxy.ProxyGenerationOptions) [0x00009] in <1da8a528a77240358e1afd90a17b1a27>:0
>   at Castle.DynamicProxy.ProxyGenerator.CreateClassProxy (System.Type,System.Type[],Castle.DynamicProxy.ProxyGenerationOptions,object[],Castle.DynamicProxy.IInterceptor[]) [0x00064] in <1da8a528a77240358e1afd90a17b1a27>:0
>   at Moq.Proxy.CastleProxyFactory.CreateProxy (System.Type,Moq.Proxy.ICallInterceptor,System.Type[],object[]) [0x0004a] in <1da8a528a77240358e1afd90a17b1a27>:0
>   at Moq.Mock`1<T_REF>.<InitializeInstance>b__24_0 () [0x00092] in <1da8a528a77240358e1afd90a17b1a27>:0
>   at Moq.PexProtector.Invoke (System.Action) [0x00001] in <1da8a528a77240358e1afd90a17b1a27>:0
>   at Moq.Mock`1<T_REF>.InitializeInstance () [0x0000c] in <1da8a528a77240358e1afd90a17b1a27>:0
>   at Moq.Mock`1<T_REF>.OnGetObject () [0x00011] in <1da8a528a77240358e1afd90a17b1a27>:0
>   at Moq.Mock.GetObject () [0x00001] in <1da8a528a77240358e1afd90a17b1a27>:0
>   at Moq.Mock.get_Object () [0x00001] in <1da8a528a77240358e1afd90a17b1a27>:0
>   at Moq.Mock`1<T_REF>.get_Object () [0x00001] in <1da8a528a77240358e1afd90a17b1a27>:0

**Comment by stakx** [comment]
This works fine on the CLR. What version of Mono are you using?

**Comment by stakx** [comment]
@mohand1993: Without any more info, it's hard to follow up on this. I suspect this is more of a problem with the Mono runtime than it is with Moq (see above) so perhaps it would be good if you reported that with the Mono team.

I'm closing this for the time being due to a lack of sufficient information. Please post back if you want to follow up on this, we can reopen the issue.

**Comment by bartdkmediahuis** [comment]
@stakx : Having the same issue :

`Mono JIT compiler version 5.4.0.201 (2017-06/71277e78f6e Thu Sep 21 19:22:55 EDT 2017)
Copyright (C) 2002-2014 Novell, Inc, Xamarin Inc and Contributors. www.mono-project.com
	TLS:           normal
	SIGSEGV:       altstack
	Notification:  kqueue
	Architecture:  amd64
	Disabled:      none
	Misc:          softdebug 
	LLVM:          yes(3.6.0svn-mono-master/8b1520c8aae)
	GC:            sgen (concurrent by default)
`

**Comment by stakx** [comment]
@mohand1993, @bartdkmediahuis: Apparently that Mono version is affected by some sort of bug, it's nothing to do with Moq. See #481. I suggest you downgrade to Mono 5.2 for the time being.
